### PR TITLE
UI option to turn on backtrace (RUST_BACKTRACE=1) --- off by default; (and also embed it in shared urls)

### PR DESCRIPTION
**Pull request by respeccing**
**UPDATE:** the following is old info, jump to <del>[updated](https://github.com/rust-lang/rust-playpen/pull/191#issuecomment-202881134)</del> <del>[updated2](https://github.com/rust-lang/rust-playpen/pull/191#issuecomment-202989809)</del> [updated3](https://github.com/rust-lang/rust-playpen/pull/191#issuecomment-203610192)

set RUST_BACKTRACE=1 when Debug build and does not set it when <s>Stable</s>Release build.

This addresses #119 although not entirely and I only say that because I imagine that a new UI
button would be desirable instead of it currently depending on the Mode(Debug/<s>Stable</s>Release) one.

Currently, if the Mode is set to Debug then the (dash)environmnet var `RUST_BACKTRACE=1` is set both for when running `rustc` (in case the compiler itself panics) and when running the output binary (`./out`). But when Mode is set to <s>Stable</s>Release, it doesn't set this env. var.

Unless I'm missing something, this should work!   as I have tested it inside a virtualbox (Manjaro(based on Archlinux) VM)

In `web.js` I've set `backtrace` to `2` which means that it will be set according to `Mode` (as explained above), but if an UI button would be added in the future, this button can directly set `backtrace` to `1` to enable backtrace or to `0` to disable it. (2 is the "auto" version, so to speak)

Thank you.

**EDIT:** Just so it's visually clear what I meant by Mode(Debug/<s>Stable</s>Release) see the red arrow in this screenshot:
debug

**EDIT2:** And by "Stable" I meant "Release"(in the UI there), for some reason in my mind Release mapped to Stable and I remembered it this way.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This was [proposed before](https://github.com/rust-lang/rust-playpen/pull/179), but I [had some reservations](https://github.com/rust-lang/rust-playpen/pull/179#issuecomment-187834724) because the spew isn't necessarily always wanted

**Comment by respeccing** [comment]
Oh I see. Since `Mode: Debug` is selected by default(ie. whenever visiting the rust-playpen website) then it's 
> always

spewed, as you've said.

Here's what I propose:

1. A separate independent setting for backtrace(on/off) with the default of `off` as to not change the default expected behavior of rust-playpen thus far (imagine old urls(that link to play.rust-lang.org expecting to see the program output without RUST_BACKTRACE being set). 

2. This setting would perhaps best be buried in Settings below `Asm Flavor:`, rather than be in plain sight near Mode: Debug/Release for example.

3. The setting ought to get embedded into the url (just like  version`(stable/beta/nightly) does) so that one can give a link to play.rust-lang.org with this setting enabled.

What do you think?

**Comment by respeccing** [comment]
I did step 1 and 2 from [my previous comment](https://github.com/rust-lang/rust-playpen/pull/191#issuecomment-202686350).

 I could do step 3 too, but this means the value for backtrace in the url would override the value in the localStorage(and then have it even be saved in the localStorage) and I'm unsure how much sense this makes ... 
 On the other hand, what if you want an url with backtrace turned on? 

 Maybe the localStorage backtrace should be replaced with the one stored in the url?(this means it would act like `version`(Stable|Beta|Nightly) aka the Channel in the screenshot)
 This way we would get both functionalities but without the persistence of the backtrace setting for subsequent rust-playpen sessions.
 But then, if it's not in localStorage does it make sense to have it under Settings? Since Settings imply they will be stored(just like others) for subsequent rust-playpen sessions(eg. next time you open browser and visit rust-playpen).

backtrace

**Comment by jonas-schievink** [comment]
I think we can set it for ASM / IR / MIR as well, maybe even unconditionally, since it will only show a backtrace on ICEs, which is always useful.

**Comment by respeccing** [comment]
> I think we can set it for ASM / IR / MIR as well, maybe even unconditionally, since it will only show a backtrace on ICEs, which is always useful.

I wouldn't mind setting it, if @alexcrichton would agree. Whether unconditionally or depending on the current setting.

**Comment by respeccing** [comment]
Alright, I've made this now so that it's satisfactorily for me.
Ignore any previous comments because, supposedly, this one summarizes them all.

Now, backtrace is a setting in Settings and it's off by default; it is also carried over in shared urls (playground URL and shortened). It is also stored in localStorage (just like `Asm Flavor` is) and it plays nicely with the setting from the url in that it only overrides it for the current session(that opened the url) - so new sessions will still use the localStorage setting.

satisfactory
(ignore the `eval=1` in the screenshot, I was just guessing around)

What's left to consider is:
1. whether to have the backtrace also for `ASM`/`LLVM IR`/`MIR` as [suggested here](https://github.com/rust-lang/rust-playpen/pull/191#issuecomment-202881996) and whether that should be unconditionally enabled or only enable it as per the backtrace setting.
2. whether the third option of `backtrace=auto` should be removed (`auto` means it turns it `on` when `Mode: Debug`, else it's `off`, eg. when `Mode: Release`) - but note that the default value for `backtrace` if `off` by default! (not `auto`)
3. any corrections/modifications/removals for what is done thus far.

Thoughts, ideas, feedback?
Thanks!

**Comment by alexcrichton** [comment]
This all sounds good to me, thanks @respeccing! I'd personally say that we should have this off by default even for the compiler because only compiler devs are really interested in the giant backtraces on ICEs, and I suspect most play users aren't compiler devs

File: bin/evaluate.sh [hunk 1]
```diff
@@ -2,6 +2,13 @@
 
 set -o errexit
 
+backtrace="--backtrace"
+if [ "${*#*$backtrace}" != "$*" ]; then
+    export RUST_BACKTRACE=1
+    set -- ${*%$backtrace*}${*#*$backtrace}
```
Comment (alexcrichton) [inline bin/evaluate.sh]: Could the caller of this script just set the environment variable for the subprocess? This seems like crazy bash syntax that only one person in the world will ever understand :(

Comment (respeccing) [inline bin/evaluate.sh]: That makes much more sense, I'll explore how to do just that. Thanks.

Comment (respeccing) [inline bin/evaluate.sh]: Oh I forgot, looks like execute() is calling playpen.execute() which doesn't allow setting any env. vars ... and this is why I had to set RUST_BACKTRACE within `evaluate.sh` depending on the incoming args (`--backtrace`)

What do you think?

Comment (respeccing) [inline bin/evaluate.sh]: **tl;dr:** I'll pass `--backtrace` as first arg, and remove it with `shift`, and also simplify the `if` test (although I borrowed the code for `if` from the below `--test` and `--color=always` code)  which results in simpler code!

-----------
Don't read this, but I'll leave it here anyway :)

I used a simpler syntax in this commit: https://github.com/rust-lang/rust-playpen/pull/191/commits/a147805ca4e75f785293bf4068025dd2c773f0e3
But that had the limitation that `--backtrace` must always be the first passed arg! else it would break things badly(eg. delete another arg!) So, if someone would modify web.py not knowing `--backtrace` must be first arg, then things would break so badly...

The ugly bash syntax(actually, it's running under the `dash` shell - so there are no arrays for example) doesn't have the above limitation and it's making sure that when `--backtrace` is specified wherever in the args list it will be safely removed, UNLESS it's specified more than once, then that will break. But I don't see why it would ever be specified more than once, unless a mistake in web.py  but still not as easy of a mistake to make than in the above case with `--backtrace` having to me first arg.

`set -- ${*%$backtrace*}${*#*$backtrace}`
This resets the `$*` args (eg. it rewrites them) to have `--backtrace` removed by concatenating the two halves of the current `$*`, eg.
if `$*` is initially `somearg --test --backtrace --color=always "arg with space" moreargs here`
and `$#` is thus `7`
The `set ` above will concat `somearg --test ` with `--color=always "arg with space" moreargs here` and set the result as being the new `$*` (the args passed to `evaluate.sh` are thus modified/reset to this new value)
and still ensure the number of args is *the same minus 1*, even if args with spaces exist!
So it becomes: `set -- somearg --test  --color=always "arg with space" moreargs here`
thus `$*` is now `somearg --test  --color=always "arg with space" moreargs here`
`$#` is `6`

Actually I forgot to test for args with spaces and it's clearly not working for that case(`$#` is `9`!)! eg. args with spaces are split into more than one arg. eg. the one argument "arg with space" becomes 3 args: "arg" "with" "space"
But I thought I did test it...

It doesn't seem to be a way to also remove `--backtrace` AND keep the args with spaces intact, and I am unsure if it's necessary to keep them intact that way...  they are only needed for this line, currently: `TERM=xterm rustc - -o ./out "$@"` In other words, if `rustc` currently(or in the future!) accepts any args with spaces(not counting the filename which is stdin here), then it's a MUST that I have to keep the args intact!  I'll investigate more...

Worst case, we could revert to that that commit, it's the simpler syntax(as you'd want) AND it keeps the args intact! Hmm... I could even just check for `$1`, expecting `--backtrace` to be the first, and only then remove it, and if it's not first, then it will cause rustc to fail because it will get `--backtrace` passed to it ... yes... this seems like a simpler way, an interestingly simple compromise between what I want and what you want:)

**Comment by respeccing** [comment]
Should I also change the **internal** values for `backtrace` that is: `on` and `off` (and `auto`) to `1`, `0` (and `2`) ?
Maybe this is desired to reduce the length of the url (since the code is limited to 5000 chars in url, apparently). And maybe while at it, I could change `backtrace` in url to just `b`.
(The UI will still show on/off/auto though.)

**Comment by respeccing** [comment]
Should I keep the variable name `backtrace=` ? I'm thinking for clarity. This affects the length of the url, for example:
http://127.0.0.1/?code=fn%20main%28%29%20{%20panic!%28%2211212%22%29%3B%20}&version=nightly&backtrace=1

I notice that `&version=nightly` is already long, but it's clear! So maybe I should keep the word `backtrace` in `&backtrace=1`  ? for clarity. Else, I could just use `bt=1` or `b=1` even, to keep the url as short as possible so that more code can get in, at the expense of clarity I guess(because some might wonder what is that `b=` in the url?!). If we later decide to change from `backtrace=` to `b=` then I imagine we'd have to support both variants for compatibility with earlier shared urls.

I would really like some feedback on this, whenever possible. Thx.

**Comment by alexcrichton** [comment]
Ah yeah I wouldn't worry about the length of the URL option, we can always change it in the future if we need to and otherewise it's probably the least of our concerns in URL length :)

File: web.py [hunk 2]
```diff
@@ -97,14 +110,27 @@ def format(version):
...
+        debug = False
+    if "2" == backtrace:
+        if debug:
+            backtrace="1"
+        else:
+            backtrace="0"
+    if "1" == backtrace:
+        args.insert(0, "--backtrace")
+        #XXX: if exists, --backtrace must be the first arg passed to compile.sh
+    if debug:
```
Comment (alexcrichton) [inline web.py]: I wonder, can this and the above block share code for building up the arguments to the compiler?

Comment (respeccing) [inline web.py]: Done(and tested). Thanks for saying that! [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is my friend ;-) (so to speak)

**Comment by respeccing** [comment]
A little summary of what has been done in this PR: (conciseness fail by the way, haha, sorry)
First, here's a screenshot(I followed a link and then I pressed Gist, then looked under Settings):
first
(in the above, the backtrace setting was not yet saved in localStorage because it was my first time visit(after clearing browser cache/stuff) and I didn't click to change it yet - only clicked to look at the setting))

The backtrace setting is `off` by default, but once you set it `on` (or `auto` which means if `Mode: Debug` then backtrace is `on`, otherwise it's `off` eg. for `Mode: Release`) it will be stored in your browser ([local storage](https://developer.mozilla.org/en-US/docs/Tools/Storage_Inspector#Local_storage_Session_storage)) and thus remembered on your subsequent visits to the website. Another such setting stored in the browser is `Asm Flavor`, so nothing new there ;)
 Shared urls(via `Shorten` and `Gist` buttons) will include the setting for backtrace in their url (except for the `Gist URL`). Because of this, shared urls that you follow will override the localStorage setting, temporarily for that session, because of the `&backtrace=val` in the url. (the `val` can be `0`, `1` or `2` for: off, on or auto, respectively). Unless you manually go and (re)set Backtrace yourself under Settings(top right wheel within the page), your previously-set (localStorage)value will remain as it were and you can see this by opening a new tab and visiting the website anew (or just removing the `&backtrace=val` from the current url and reload the page by pressing Enter, for example.) then look at it under Settings.

The backtrace setting does not affect `rustfmt` (the `Format` button) so if this one gets an ICE, you won't see it in the output. (I could implement this but would require creating an extra (wrapper) script, similar to `evaluate.sh`, for running only `rustfmt` - which I'd be happy to do, if this is an acceptable idea; the script is needed in order to be able to pass an arg to it, so it then can set the env. var(RUST_BACKTRACE=1), because `rustfmt` too is ran directly by `playpen`(which is ran by `web.py`) and playpen has no real way of passing outer env. vars, or setting them(well, aside from `HOSTNAME` I guess, which it accepts as arg.))

The backtrace setting does affect the following actions: `Run`(/`Test`), `ASM`, `LLVM IR`, `MIR`, `Shorten`, `Gist`, and if Backtrace is set to `auto`, it affects `Mode: Debug`(turns on backtrace) and `Mode: Release`(turns off backtrace). 
(I can remove this extra option of `auto`, before this PR is merged, if this seems extraneous or isn't wanted; it's really not a problem! I won't feel like I've wasted anything if it's removed, btw; just somebody say so, if it's so desired!)

File: bin/evaluate.sh (hunk 1 above)
Comment (respeccing) [inline bin/evaluate.sh]: Hmm, I just realized something... the caller(`webpy`) can actually call `playpen` with the shell(`dash`, not bash) and pass it the script, but the best part is that, this way it could also set the env var, like so:
`playpen ppargshere -- /usr/bin/dash -c 'export RUST_BACKTRACE=1; evaluate.sh`
or
`playpen ppargshere -- /usr/bin/dash -c 'RUST_BACKTRACE=1 evaluate.sh`

(this I mentioned like 1 day ago in another issue and I forgot about it completely, see "EDIT3" [here](https://github.com/rust-lang/rust-playpen/issues/119#issuecomment-202261425) for the source of this)

So the question is, should I change this PR to do this?! it seems much better than having to pass(then remove) the `--backtrace` arg!

**Comment by alexcrichton** [comment]
Using extra shell syntax to specify the RUST_BACKTRACE env var actually seems pretty reasonable to me, could that be implemented here as well?

**Comment by respeccing** [comment]
Really enjoyed this thus far. Don't hesitate to tell me if any changes are required or anything like that.
(meanwhile I'll recheck and see if I forgot something)

File: web.py [hunk 3]
```diff
@@ -57,22 +57,39 @@ def wrapper(*args, **kwargs):
...
+    else:
+        debug = False
+    if "2" == backtrace:
+        if debug:
+            backtrace="1"
+        else:
+            backtrace="0"
+    if "1" == backtrace:
+        env_vars.append("RUST_BACKTRACE=1")
+    if debug:
```
Comment (alexcrichton) [inline web.py]: Perhaps this could just be `if optimize == "0"`?

File: web.py [hunk 4]
```diff
@@ -57,22 +57,39 @@ def wrapper(*args, **kwargs):
...
+    if "0" == optimize: #this means Debug, else Release
+        debug = True
+    else:
+        debug = False
+    if "2" == backtrace:
+        if debug:
+            backtrace="1"
+        else:
+            backtrace="0"
+    if "1" == backtrace:
```
Comment (alexcrichton) [inline web.py]: could this just be `if backtrace == "2" or debug`?

Comment (alexcrichton) [inline web.py]: er, `or optimize == "0"`*

File: playpen.py [hunk 5]
```diff
@@ -1,8 +1,63 @@
 #!/usr/bin/env python3
 
 import subprocess
+import shlex #for quote() - python3 required
+import re #for re.match()
 
-def execute(version, command, arguments, data=None):
+def execute(version, command, arguments, data=None, env_vars=None):
+    #Note: arguments and env_vars are expected to be of tuple type, not list,
```
Comment (alexcrichton) [inline playpen.py]: Oh wow! This actually looks like it's a bit noisy, could the implementation be simplified? We don't really need to do any validation of the command line or anything like that as we're just spawning processes ourselves

Comment (respeccing) [inline playpen.py]: **tl;dr4**: ok, don't read any of this, you'll just see what I mean in the (future)commit - and if you really want me to avoid escaping args(which is done ONLY when backtrace is on) then, fine :)

**tl;dr3:** I can reduce this logic as you want it, but to ensure consistency of subprocess execution between when backtrace is on and when it's off, I have to quote the `arguments` passed to `command` but only in the former case(of backtrace being on). Else, they are still consistent(for our known args thus far) but only until any one arg(added by someone in the future) contains a bad char(space, quotes, backticks, dollar signs, etc.)

**tl;dr2:** the way we decided to pass the env. var RUST_BACKTRACE(via `dash -c "export RUST_BACKTRACE=1; command args"`), requires to escape/quote the args passed to `command`(eg. to `evaluate.sh`) in order to maintain the same logic for when backtrace is off(ie. executing just `command args`, where the args are each elements of a tuple when passed to `subprocess.Popen`, whereas with the `dash -c` logic the args and command are both just one element of a tuple like so: `("/usr/bin/dash", "-c", "command args")) - ok i'm bad at tl;dr-s

**tl;dr:** Should I escape/quote the args when backtrace is on? or else the logic for passing args to command but-only-when-backtrace-is-on is in danger of breaking-but-only-in-the-case-when future new args would contain any spaces, dollar signs, single or double quotes, backticks etc.

**I can definitely reduce this as you want it**, but one thing I still have to do is to quote the args(but only in the case when backtrace is on, because this is the only time when we wrap the `command` within the `dash` shell), or else `dash` won't have the same guarantee that just running the `command`(without `dash` wrapped around it, that is) *currently* has, which is: an arg for example can contain spaces, and it would be seen as one arg not two; to ensure this is true even with `dash -c "command args"` I must escape the args (even though `command args` is sent as one argument to the `dash` command, it's the `command` that needs to see each of the args as whole as they are like when sent directly through `command args` without the `dash` wrapped around)

For example:
`evaluate.sh "this is one arg"`
vs.
`dash "-c" "evaluate.sh this is one arg"`
(in the above, I added double quotes for visibility only! to signify the arg boundaries)
So `this is one arg` becomes 4 args in the latter case! Unless I escape it (with shlex.quote()), in which case it becomes:
`dash "-c" "evaluate.sh 'this is one arg'"`
(with single quotes present in string; to reiterate, the double quotes are not present in string - they're just for visibility here)

You could argue that since we do know what the args are right NOW in this moment of time(`--test`, `--color=always`, `-C`, `opt-level=0`, `-g`, maybe others), we do not need to escape(quote) any! But who would remember to look in web.py's `execute` function (which is where I'll move the overly simplified specific logic for this - I have to, because it's no longer generic to stay here in `playpen.py`) and to make sure the supposedly newly passed args(of the future) are not split into two args because they now contain spaces(or backticks or quotes, or dollar signs).

**Comment by respeccing** [comment]
Done as requested ;-)
